feat(otlp-exporter): add timeout env var

svetlanabrennan opened the pull request:
Signed-off-by: Svetlana Brennan <svetlana.svn@gmail.com>

<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/master/CONTRIBUTING.md#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

Partially Fixes #2706 (issue)

This is just the first PR for issue #2706 that adds `OTEL_EXPORTER_OTLP_TIMEOUT` & `OTEL_EXPORTER_OTLP_TRACES_TIMEOUT` as an environment variable for the otlp trace exporters. I wanted to split this issue up in separate prs so it's easier to review and I need this timeout variable for [another pr regarding retry logic](https://github.com/open-telemetry/opentelemetry-js/pull/2717) I'm working on. 

## Short description of the changes

- Added `OTEL_EXPORTER_OTLP_TIMEOUT` as default environment variable
- Added `OTEL_EXPORTER_OTLP_TRACES_TIMEOUT` as default environment variable
- Added `timeoutMillis` property to the `OTLPExporterBase` class
- Added timeout logic to http request (`sendWithHttp` utility function)
- Added timeout logic to xhr request (`sendWithXhr` utility function)
- Added timeout logic to grpc request (`send` utility function)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Unit tests
- [x] Integration tests

## Checklist:

- [x] Followed the style guidelines of this project
- [x] Unit tests have been added
- [x] Documentation has been updated

svetlanabrennan commented:
I haven't updated documentation yet. I would like some feedback on this pr before I update documentation please.

svetlanabrennan commented:
Does `OTEL_EXPORTER_OTLP_TRACES_TIMEOUT` take precedence over `OTEL_EXPORTER_OTLP_TIMEOUT`?

vmarchaud commented:
> oes OTEL_EXPORTER_OTLP_TRACES_TIMEOUT take precedence over OTEL_EXPORTER_OTLP_TIMEOUT

Yes it should, signal specific configs take priority over generic ones

svetlanabrennan commented:
Got it. I just implemented that logic for `OTEL_EXPORTER_OTLP_TRACES_TIMEOUT` to take precedence over `OTEL_EXPORTER_OTLP_TIMEOUT` and refactored some code.

vmarchaud commented on packages/exporter-trace-otlp-http/src/OTLPExporterBase.ts:
```diff
@@ -102,6 +110,26 @@ export abstract class OTLPExporterBase<
+            : defaultTimeout;
+  }
+
+  private _invalidTimeout(timeout: number, defaultTimeout: number): number {
```
i think you should inline this check inside the above config

svetlanabrennan replied:
This is fixed.

vmarchaud commented on packages/exporter-trace-otlp-http/src/OTLPExporterBase.ts:
```diff
@@ -102,6 +110,26 @@ export abstract class OTLPExporterBase<
+  private _configureTimeout(timeoutMillis: number | undefined): number {
+    let defaultTimeout = getEnv().OTEL_EXPORTER_OTLP_TRACES_TIMEOUT;
+
+    if ( defaultTimeout === null) {
```
the default cannot be null right ? its 10000 by default

svetlanabrennan replied:
That's correct. So if `OTEL_EXPORTER_OTLP_TRACES_TIMEOUT` takes priority over the generic timeout `OTEL_EXPORTER_OTLP_TIMEOUT` environment variable, would it make sense to just take out the mention of `OTEL_EXPORTER_OTLP_TIMEOUT` here? 

I'm having a hard time understanding the use of `OTEL_EXPORTER_OTLP_TIMEOUT` in the otlp trace exporters?

vmarchaud replied:
I think we are meant to check if `OTEL_EXPORTER_OTLP_TRACES_TIMEOUT` is defined, if so yes use it then check `OTEL_EXPORTER_OTLP_TIMEOUT` and finally fallback to the default value

svetlanabrennan replied:
Got it. 

I added some logic around this selection process [here](https://github.com/open-telemetry/opentelemetry-js/blob/da3211f44bf32d4568fe6ceee660b4996d897eed/packages/exporter-trace-otlp-http/src/OTLPExporterBase.ts#L123). 

If a user defined `OTEL_EXPORTER_OTLP_TRACES_TIMEOUT` then it will use that first. 

Otherwise it'll check if the user defined `OTEL_EXPORTER_OTLP_TIMEOUT` and use that instead. 

And if neither of those are defined then it'll use the default `OTEL_EXPORTER_OTLP_TRACES_TIMEOUT` as specified in the exporter environments which is 10000.

svetlanabrennan replied:
Another question:

Is the exporter timeout allowed to be set via config settings when creating a new `OTLPTraceExporter`?

Example: 

```js
const collectorOptions = {
  timeoutMillis: 15000
};

const traceExporter = new OTLPTraceExporter(collectorOptions);
```

That is how I have it setup now that the exporter will use `timeoutMillis` in the `collectorOptions` if it's provided. If there is no `timeoutMillis` provided in the `collectorOptions` then the exporter will check the user defined environment variables (`OTEL_EXPORTER_OTLP_TRACES_TIMEOUT` and `OTEL_EXPORTER_OTLP_TIMEOUT`).

I assumed that if the `BatchSpanProcessor` timeout is set that way (through the config settings), then it would be a similar process for setting timeout in the otlp trace exporter. But please correct me if I'm wrong.

vmarchaud replied:
I guess if thats not spec'ed its up to us, i agree that we should either be able to control options via the environment or the code

svetlanabrennan replied:
Sounds good. I'll add some information to the readme for each otlp trace exporter about setting this timeout via code.

vmarchaud commented on packages/exporter-trace-otlp-http/src/OTLPExporterBase.ts:
```diff
@@ -93,6 +96,11 @@ export abstract class OTLPExporterBase<
     return new Promise<void>((resolve, reject) => {
+      setTimeout(() => {
+        // don't wait anymore for export
+        reject(new Error('Timeout'));
```
do this will be logged in the diag level by the sdk ?

svetlanabrennan replied:
I'm not that familiar with the diag level logger but I've ran some tests locally with a sample app and when I have this set: 

```
api.diag.setLogger(
  new api.DiagConsoleLogger(),
  api.DiagLogLevel.ERROR
);
```

then I get a `{"stack":"Error: Export Timeout\n    at Timeout._onTimeout `....etc error log in the console. 

Is that what you were asking about?

dyladan replied:
Yes these errors are caught. I think @vmarchaud just wanted to make sure we weren't going to cause a process crash or unhandled promise rejection or anything like that.

svetlanabrennan commented:
Added timeout config information to each otlp trace exporter readme. 

This pr is ready for review/merge from my end.

svetlanabrennan commented on packages/exporter-trace-otlp-http/src/OTLPExporterBase.ts:
```diff
@@ -111,21 +111,23 @@ export abstract class OTLPExporterBase<
     return typeof timeoutMillis === 'number'
       ? timeoutMillis < 0
-        ? this._invalidTimeout(timeoutMillis, timeout)
+        ? this._invalidTimeout(timeoutMillis, 10000)
```
Done

blumamir commented on packages/exporter-trace-otlp-http/src/OTLPExporterBase.ts:
```diff
@@ -93,6 +98,11 @@ export abstract class OTLPExporterBase<
 
   private _export(items: ExportItem[]): Promise<unknown> {
     return new Promise<void>((resolve, reject) => {
+      setTimeout(() => {
```
we need to `unref` this timer so it does not hold the process alive in case of shutdown

svetlanabrennan replied:
Done.

blumamir commented on packages/exporter-trace-otlp-http/src/OTLPExporterBase.ts:
```diff
@@ -102,6 +112,32 @@ export abstract class OTLPExporterBase<
     });
   }
 
+  private _configureTimeout(timeoutMillis: number | undefined): number {
```
is it feasible to write any unit test for this logic?

svetlanabrennan replied:
My thought was that since the method is private, we couldn't test it. I did test the logic locally in one of my sample apps.

blumamir replied:
Yeah, it's not very easy.

Since these functions are private methods in the class, it's a bit complicated. They can actually be static functions (they don't access the `this` parameter) and maybe be moved to utils. This way it's trivial to test them, and I think it's common in other parts of the project to have these kinds of functions as `utils`.

Another alternative, which will also validate the timeout logic itself, would be to write tests similar to the existing ones, with `fakeRequest` that takes long time to reply and let a short (1 ms) timeout hit. Then we can verify the promise is rejected with the proper message.

svetlanabrennan replied:
Ok. I see what you mean by moving those methods to `utils` and testing it. I will do that. 

And I will test the timeout logic with the `fakeRequest` test.

svetlanabrennan replied:
I moved the functions around configuring timeout to the `utils` and added unit testing for that logic [here](https://github.com/open-telemetry/opentelemetry-js/blob/88660355303012cda492b9842492516d5369b33a/packages/exporter-trace-otlp-http/test/common/utils.test.ts#L55).

blumamir commented on packages/exporter-trace-otlp-grpc/README.md:
```diff
@@ -107,6 +107,29 @@ provider.register();
 
+## Exporter Timeout Configuration
+
+The OTLPTraceExporter has a timeout configuration option which is the maximum time, in milliseconds, the OTLP exporter will wait for each batch export. The default value is 10000ms.
```
discussion:
I wonder if this time should be set in seconds or ms.
ms feels more native to JS.
On the other end, the spec states the default value as `10s` and ruby and python use seconds for this value, and there is also a value in consistency across otel SDKs.
WDYT?

svetlanabrennan replied:
I see what you mean. It's interesting that timeout for the BSP is [using ms](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md) and timeout for the exporter mentions seconds. 

I do like the consistency across other otel SDKs so I'm good with changing this to use seconds.

blumamir replied:
> It's interesting that timeout for the BSP is [using ms](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md) and timeout for the exporter mentions seconds.

Interesting. I wonder if there is any thinking behind this or it was just added at different times and there was no notice on unifying it.

> I do like the consistency across other otel SDKs so I'm good with changing this to use seconds.

I now checked the java implementation and it uses ms, so there is currently no consistency with popular otel SDKs. I am good with keeping it as ms or changing it to seconds. Whichever one you think is better. Sorry for the trouble.

svetlanabrennan replied:
No worries. I will keep it as ms.

blumamir commented on packages/exporter-trace-otlp-http/src/OTLPExporterBase.ts:
```diff
@@ -102,6 +112,32 @@ export abstract class OTLPExporterBase<
 
+  private _configureTimeout(timeoutMillis: number | undefined): number {
+    return typeof timeoutMillis === 'number'
+      ? timeoutMillis < 0
```
discussion:
I wonder what should be the semantic meaning of timeout `0`.
If the user sets the value to 0, it will always timeout before there is any chance for the export to report real status which doesn't make a lot of sense to me. 
Sometimes the value 0 for timeout is interpreted as "disable" or "no timeout at all".
Not sure what is the expected behavior here. I wish the spec stated it but looks like it does not.

svetlanabrennan replied:
I wasn't sure what to do with timeout when it's set to `0` since the spec mentions warning the user only when it's negative. But I do feel like if the user sets `0` then we should use warn them about that timeout value and use the `DEFAULT_TIMEOUT` value (10000ms) instead.

blumamir replied:
If we do decide to forward the timeout to the `http` package, then we can just fallback to it's implementation.
I couldn't find how timeout of 0 is handled, but I guess it's being forwarded to the os [where timeout 0 means "no timeout"](https://linux.die.net/man/3/setsockopt):

> SO_RCVTIMEO
> Sets the timeout value that specifies the maximum amount of time an input function waits until it completes. It accepts a timeval structure with the number of seconds and microseconds specifying the limit on how long to wait for an input operation to complete. If a receive operation has blocked for this much time without receiving additional data, it shall return with a partial count or errno set to [EAGAIN] or [EWOULDBLOCK] if no data is received. **The default for this option is zero, which indicates that a receive operation shall not time out**. This option takes a timeval structure. Note that not all implementations allow this option to be set.

Anyway, it's a minor detail, so feel free to implement whatever you think is right and resolve the conversation when you are settled :)

svetlanabrennan replied:
I went with checking if timeout of 0 is set by the user and if it is, then just display the warning and use default timeout as specified in the environment.ts file.

svetlanabrennan commented on packages/exporter-trace-otlp-http/src/OTLPExporterBase.ts:
```diff
@@ -102,6 +112,32 @@ export abstract class OTLPExporterBase<
+          : this._selectATimeout();
+  }
+
+  private _selectATimeout(): number {
```
I agree. I'll make that change.

svetlanabrennan commented on packages/exporter-trace-otlp-http/src/OTLPExporterBase.ts:
```diff
@@ -102,6 +112,32 @@ export abstract class OTLPExporterBase<
   }
 
+  private _configureTimeout(timeoutMillis: number | undefined): number {
+    return typeof timeoutMillis === 'number'
+      ? timeoutMillis < 0
+        ? this._invalidTimeout(timeoutMillis, DEFAULT_TIMEOUT)
+        : timeoutMillis
+          : this._selectATimeout();
```
That makes sense. I'll update it to use if statements.

svetlanabrennan commented on packages/opentelemetry-core/test/utils/environment.test.ts:
```diff
@@ -92,6 +92,8 @@ describe('environment', () => {
         OTEL_SPAN_LINK_COUNT_LIMIT: 30,
         OTEL_TRACES_SAMPLER: 'always_on',
         OTEL_TRACES_SAMPLER_ARG: '0.5',
+        OTEL_EXPORTER_OTLP_TIMEOUT: 10000,
+        OTEL_EXPORTER_OTLP_TRACES_TIMEOUT: 10000,
```
Got it. I'll make that change.

svetlanabrennan commented on packages/exporter-trace-otlp-http/src/OTLPExporterBase.ts:
```diff
@@ -102,6 +112,32 @@ export abstract class OTLPExporterBase<
+  }
+
+  private _invalidTimeout(timeout: number, defaultTimeout: number): number {
+    diag.warn('Timeout must be non-negative', timeout);
```
Done.

blumamir commented:
I read again the python, ruby and java implementation of this feature, and they all forward the timeout parameter to the underlying http / grpc request which I assume use it for the actual timeout on the os socket.
References:
[python grpc](https://github.com/open-telemetry/opentelemetry-python/blob/d71cf688d466186d9597d3466b22239a647d20dd/exporter/opentelemetry-exporter-otlp-proto-grpc/src/opentelemetry/exporter/otlp/proto/grpc/exporter.py#L299)
[python http](https://github.com/open-telemetry/opentelemetry-python/blob/d71cf688d466186d9597d3466b22239a647d20dd/exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/trace_exporter/__init__.py#L111)
[ruby http](https://github.com/open-telemetry/opentelemetry-ruby/blob/ff3a5491bdc05262e9add4449a72d1a2244cf96b/exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb#L169)
[java http](https://github.com/open-telemetry/opentelemetry-java/blob/a047632ad75255ecf8a7ac8d783be842f101651e/exporters/otlp-http/trace/src/main/java/io/opentelemetry/exporter/otlp/http/trace/OtlpHttpSpanExporterBuilder.java#L35)

I think there are the following benefits in taking this approach over setting and applicative timeout timer:
- If the timeout is configured as a large value (larger than the OS default timeout), then the request will be actually given the chance to complete within this requested timeframe and will not terminate prematurely by the OS.
- If the timeout is smaller than the OS's default, then we will release OS resources (socket) as early as it is no longer needed, which introduces performance imporvment.
- The implementation becomes a bit simpler as we only need to forward the value instead of maintaining timers in the exporter codebase.

timeout support:
[nodejs http](https://nodejs.org/api/http.html#requestsettimeouttimeout-callback)
[nodejs grpc](https://stackify.dev/151544-setting-timeouts-for-grpc-functions-in-node-js)
[web XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/timeout)

I am also good with keeping the current implementation if you prefer.

blumamir commented on packages/exporter-trace-otlp-http/src/OTLPExporterBase.ts:
```diff
@@ -102,6 +112,32 @@ export abstract class OTLPExporterBase<
+      ? definedTimeout < 0
+        ? this._invalidTimeout(definedTimeout, DEFAULT_TIMEOUT)
+        : definedTimeout
+          : getEnv().OTEL_EXPORTER_OTLP_TRACES_TIMEOUT;
```
I think the idle solution is to have just a single default in environment and use it everywhere.
But I see how it's not so trivial to use in this case.

For reference: [java is throwing an exception](https://github.com/open-telemetry/opentelemetry-java/blob/a047632ad75255ecf8a7ac8d783be842f101651e/exporters/otlp-http/trace/src/main/java/io/opentelemetry/exporter/otlp/http/trace/OtlpHttpSpanExporterBuilder.java#L34) if the timeout is negative, preventing the exporter from being initialized.

svetlanabrennan replied:
I see. I went with just providing a warning if the value is negative (and in this case I will add if it's set to 0) because the [spec says to ignore that settings and use the default in that case](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#duration):

svetlanabrennan commented:
My plan was to use the timeout in the `sendWithHttp` util method in my issue for adding http retry functionality as soon as this pr gets merged and that method sends the request. Which would mean moving the setTimeout around this timeout in this pr to that `sendWithHttp` method. 

But I can move the timeout functionality to `sendWithHttp` now instead of waiting to move it in my retry logic pr.

svetlanabrennan commented:
@blumamir Here's an update on the request timeout chat we had:

I moved the logic around using the export timeout to cancel exporting spans from `src/OTLPExporterBase.ts` to [`sendWithHttp` in util.js](https://github.com/open-telemetry/opentelemetry-js/blob/88660355303012cda492b9842492516d5369b33a/packages/exporter-trace-otlp-http/src/platform/node/util.ts#L80) and to [`sendWithXhr` in util.js](https://github.com/open-telemetry/opentelemetry-js/blob/88660355303012cda492b9842492516d5369b33a/packages/exporter-trace-otlp-http/src/platform/browser/util.ts#L66). 

There is no default timeout for an http request so I chose to go with [`req.setTimeout` and use the exporter timeout value here. ](https://github.com/open-telemetry/opentelemetry-js/blob/88660355303012cda492b9842492516d5369b33a/packages/exporter-trace-otlp-http/src/platform/node/util.ts#L80). This way I can use this logic in my http retry pr and in this http retry pr I will set `options.timeout` to a low value (ex. 1 sec) so there is enough time to retry the request several times within the allowed exporter timeout value (ex. 10 secs). I can't set `options.timeout` value to the exporter timeout value because there won't be enough time to retry since the 1st http request will use up all the allowed exporter timeout value. 

I hope that explanation makes sense but let me know if I can clarify further.

svetlanabrennan commented:
Question on http and xhr request tests. 

1. Http request tests:

After I added `req.SetTimeout` to the http requests, some existing tests were failing [here](https://github.com/open-telemetry/opentelemetry-js/blob/88660355303012cda492b9842492516d5369b33a/packages/exporter-trace-otlp-http/test/node/CollectorTraceExporter.test.ts#L120) around `stubRequest` with error `req.setTimeout is not a function` which was originating from `PassThrough`. 

I added `sinon.spy()` on `fakeRequest.setTimeout` [here](https://github.com/open-telemetry/opentelemetry-js/blob/88660355303012cda492b9842492516d5369b33a/packages/exporter-trace-otlp-http/test/node/CollectorTraceExporter.test.ts#L120) which seemed to fix the error. Is this the recommended approach to take to solve this `req.setTimeout is not a function` issue.

2. I noticed that [browser tests are excluded](https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/exporter-trace-otlp-http/package.json#L24) so I didn't write any test for `xhr.timeout` yet. Should I still write tests for it even though browser tests are excluded?

Flarna commented on packages/exporter-trace-otlp-http/src/platform/browser/util.ts:
```diff
@@ -62,13 +63,19 @@ export function sendWithXhr(
   });
 
+  xhr.ontimeout = function () {
+    xhr.abort();
```
is this needed? I'm not an browser expert but reading following at MDN:
```
XMLHttpRequest.timeout
Is an unsigned long representing the number of milliseconds a request can take before automatically being terminated.
```
I interpret this that abort happens automatic and the `ontimeout` handler is just a FYI to log/improve error code,...

Flarna commented on packages/exporter-trace-otlp-http/src/platform/browser/util.ts:
```diff
@@ -52,7 +52,8 @@ export function sendWithXhr(
   onSuccess: () => void,
-  onError: (error: otlpTypes.OTLPExporterError) => void
+  onError: (error: otlpTypes.OTLPExporterError) => void,
+  exporterTimeout: number
```
usually callbacks are last but not sure how often this API is used and if it is exported to public. So feel free to keep as is.

dyladan replied:
it is not public

Flarna commented on packages/exporter-trace-otlp-http/src/platform/node/util.ts:
```diff
@@ -76,9 +77,16 @@ export function sendWithHttp<ExportItem, ServiceRequest>(
     });
   });
 
+  req.setTimeout(exporterTimeout, function requestTimeout () {
```
I don't think that this does actually what is expected there for following reasons:
* according to [HTTP docs](https://nodejs.org/dist/latest-v16.x/docs/api/http.html#requestsettimeouttimeout-callback) the time until a socket is assigned is not included here
* once a socket is assigned it measures socket inactivity according to [NET docs](https://nodejs.org/dist/latest-v16.x/docs/api/net.html#socketsettimeouttimeout-callback)

So if a connection is terrible slow but alive no timeout will happen.

This is different to the timeout option in XHR which specifies how long a request may take.

I think we need a dedicated timer here instead of using the `setTimeout` API from HTTP.

svetlanabrennan replied:
I can replace with a general timer that uses the exporter timeout:

```js 
setTimeout(() => { 
  // don't wait anymore for export 
  onError(new Error('Timeout')); 
}, exporterTimeout); 
```
    
And then set a timeout on the req with `req.setTimeout` to handle request timeouts and set [`options.timeout`](https://nodejs.org/api/http.html#httprequesturl-options-callback) which from my understanding will set the timeout before a socket is connected.

Flarna replied:
`options.timeout` and [agent.option.timeout](https://nodejs.org/dist/latest-v16.x/docs/api/http.html#new-agentoptions) are actually just other ways to call `request.timeout()` - it's done internally.

But to my understanding this timeout measures socket inactivity, not connection setup time, time for sending body (on slow connections,...),...

I think the correct approach is to just start a timer and abort/destroy the outgoing request. Clearly the time should be stopped once the request finished.

Flarna replied:
I did a quick look into java. They seem to [use](https://github.com/open-telemetry/opentelemetry-java/blob/87f788a3b46fc124910f77a516358319cc0ee75f/exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/okhttp/OkHttpExporterBuilder.java#L97-L99) the `OkHttpClient` which offers a [`callTimeout`](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-ok-http-client/-builder/call-timeout/) API. The timeout given there is a timeout for the complete request, similar as in XHR.

in Node.js `request.setTimeout` is delegated to `socket.setTimeout` which is a socket inactivity timeout. It may fire as expected in some cases but in others it doesn't.

Here a sample where it never fires because socket is continously doing something:
```js
const http = require("http");

const start = Date.now();
const port = 8000;

function log(...args) {
    console.log(Date.now() - start, ...args);
}

http.createServer((req, res) => {
    log("request start");
    const i = setInterval(() => {
        log("read");
        const data = req.read(100);
        if (!data) {
            clearInterval(i);
            res.end();
        }
    }, 100)
}).listen(port, "localhost");

const req = http.request({ port, timeout: 3000, method: "POST" }, (res) => {
    log("response received");
});
req.write(Buffer.alloc(1000));
setInterval(() => req.write(Buffer.alloc(1000)), 1000);
req.on("timeout", () => log("http Timeout"));
```

If the way how exporter uses http can ensure that such thingy is never happening then fine.

svetlanabrennan replied:
Ok. So if we use a general timer and destroy the req after a few seconds then it'll destroy the request even if the socket is still in connection setup phase and we would still have time to retry the request again?

Here's a quick mock up of how this general request timer can work with the exporter timer. Sorry for the long code example. 

```js
// example retry attempts and backoff
const DEFAULT_MAX_ATTEMPTS = 4;
const DEFAULT_INITIAL_BACKOFF = 1000;
const DEFAULT_BACKOFF_MULTIPLIER = 1.5;

function sendWithHttp(collector, data, contentType, onSuccess, onError) {
    let retryTimer;

   // reject promise since exporter had 10 secs to export
   // clear retryTimer to prevent possible queued retry requests
    let exporterTimer = setTimeout(() => {
        clearTimeout(retryTimer);
        onError(new Error('Request Timeout'));
    }, collector._timeoutMillis);

    const parsedUrl = new url.URL(collector.url);
    const options = {
        hostname: parsedUrl.hostname,
        port: parsedUrl.port,
        path: parsedUrl.pathname,
        method: 'POST',
        headers: Object.assign({ 'Content-Length': Buffer.byteLength(data), 'Content-Type': contentType }, collector.headers),
        agent: collector.agent
    };

    const request = parsedUrl.protocol === 'http:' ? http.request : https.request;

    const sendWithRetry = (retries = DEFAULT_MAX_ATTEMPTS, backoffMillis = DEFAULT_INITIAL_BACKOFF) => {
        let reqIsDestroyed;

        // general 2 seconds timer for each request
        // req.destroy() will emit an error
        let oneRequestTimer = setTimeout(() => {
            reqIsDestroyed = true;
            req.destroy();
        }, 2000);

        const req = request(options, (res) => {
            let responseData = '';
            res.on('data', chunk => (responseData += chunk));
            res.on('end', () => {
    
                if (res.statusCode && res.statusCode < 299) {
                    api_1.diag.debug(`statusCode: ${res.statusCode}`, responseData);
                    
                    // clear all timers since request was successful and resolve promise
                    clearTimeout(oneRequestTimer);
                    clearTimeout(exporterTimer);
                    clearTimeout(retryTimer);
                    onSuccess();
                } else if (isRetryable(res.statusCode) && retries > 0) {
                    // clear request 2 second timer and try request again
                    clearTimeout(oneRequestTimer);

                    retryTimer = setTimeout(() => {
                        // try request again with backoff
                        return sendWithRetry(retries - 1, backoffMillis * DEFAULT_BACKOFF_MULTIPLIER);
                    }, backoffMillis)
                } else {
                    const error = new otlpTypes.OTLPExporterError(
                        res.statusMessage,
                        res.statusCode,
                        responseData
                    );
                    
                    // clear all timers since request failed and there are no more retries left
                    // then reject promise
                    clearTimeout(oneRequestTimer);
                    clearTimeout(exporterTimer);
                    clearTimeout(retryTimer);
                    onError(error);
                }
            });
        });
    
        req.on('error', (error) => {
            // req was destroyed due to timeout, try again if there are remaining retries
            if (error.code === 'ECONNRESET' && reqIsDestroyed && retries > 0) {
                
                retryTimer = setTimeout(() => {
                    // try request again with backoff
                    return sendWithRetry(retries - 1, backoffMillis * DEFAULT_BACKOFF_MULTIPLIER);
                }, backoffMillis)
                
             // if something else caused an error reject promise
            } else {
                onError(error);
            }
        });
    
        switch (collector.compression) {
            case types_1.CompressionAlgorithm.GZIP: {
                if (!gzip) {
                    gzip = zlib.createGzip();
                }
                req.setHeader('Content-Encoding', 'gzip');
                const dataStream = readableFromBuffer(data);
                dataStream.on('error', onError)
                    .pipe(gzip).on('error', onError)
                    .pipe(req);
                break;
            }
            default:
                req.write(data);
                req.end();
                break;
        }
    }
    sendWithRetry();
}

function isRetryable(statusCode) {
    const retryCodes = [429, 502, 503, 504];
  
    return retryCodes.includes(statusCode);
}
```

I tested this locally (so far) on a sample app with a socket hang up issue and was able to retry the request 3 times (with the current retry backoff strategy). 

Thoughts on this?

Flarna replied:
I think this could work. In general it's hard to follow node HTTP requests in all special cases and there may be subtle differences between the node versions.
It may be needed to listen also on the `error` event of response object as it seems that errors are emitted there once connection was successful but the data phase takes to long (see [here](https://nodejs.org/dist/latest-v16.x/docs/api/http.html#httprequesturl-options-callback), scroll a bit down for the use cases).

I'm also not sure if `error.code === 'ECONNRESET'` is true in all cases.

Flarna commented on packages/exporter-trace-otlp-http/src/platform/node/util.ts:
```diff
@@ -76,9 +77,16 @@ export function sendWithHttp<ExportItem, ServiceRequest>(
-  req.on('error', (error: Error) => {
-    onError(error);
+  req.on('error', (error: Error | any) => {
+    if (error.code === 'ECONNRESET') {
```
you may get `ECONNRESET` for several reasons, not just request timeout (e.g. server decides to end the connection...). see  [HTTP docs](https://nodejs.org/dist/latest-v16.x/docs/api/http.html) for some examples.

svetlanabrennan replied:
I can add an additional check to see if a [request was called with `destroy` ](https://nodejs.org/dist/latest-v16.x/docs/api/http.html#requestdestroyed) so:

```js
if (error.code === 'ECONNRESET' & req.destroyed)
```

Flarna replied:
`req.destroyed` is only available for node >=14.1.0.
But I think it should be not that hard to store this info somewhere in exporter at the time `req.destroy` is called.

svetlanabrennan replied:
Got it. I'll store it in a variable.

Flarna commented on packages/exporter-trace-otlp-grpc/README.md:
```diff
@@ -107,6 +107,29 @@ provider.register();
 
 Note, that this will only work if TLS is also configured on the server.
 
+## Exporter Timeout Configuration
```
Documentation how to configure timeout in GRPC exporter is added here but I miss any changes for the actual implementation. Is this working already but just not documented?

svetlanabrennan commented:
I wanted to provide a clarification on how the Java and .Net SDKs are using the export timeout/timer. 

They are setting the export process timeout to the 10 secs (default value as mentioned by the spec) or whatever the user sets that timeout value to but they are **not dictating the timeout of each phase of a request or a timeout for an entire single request**. They are using the defaults of their http library to dictate the timeout for each phase of the request and/or an entire request. 

So if one requests uses up all the allowed exporter time value (ex. 10 secs) to send the request and the request has not completed yet, then it'll abort the request and not retry. Or if one request fails and there's time remaining on the export timer then it'll retry the request again. 

This makes sense to me because we don't know the size of each payload so we don't know if it'll take 4 seconds for 8 seconds for the request to complete. So if a request with a large payload takes longer to complete, we should let the request do that within the allowed export time instead of aborting the request early (ex. before the 10 seconds mark).

Do you agrees with this? @dyladan @blumamir @Flarna

Flarna commented:
I'm fine with this approach.
There are a lot variants like different timeouts for connection setup, request sending, response consumption,.. which result in more complicated solutions. If it turns out the simple variant has issues it's always possible to improve/modifiy it later but I see no reason to start with the most complicated one.

dyladan commented:
> I wanted to provide some clarification on how the Java and .Net SDKs are using the export timeout/timer.
> 
> They are setting the export process timeout to the 10 secs (default value as mentioned by the spec) or whatever the user sets that timeout value to but they are **not dictating the timeout of each phase of a request or a timeout for an entire single request**. They are using the defaults of their http library to dictate the timeout for each phase of the request and/or an entire request.
> 
> So if one requests uses up all the allowed exporter time value (ex. 10 secs) to send the request and the request has not completed yet, then it'll abort the request and not retry. Or if one request fails and there's time remaining on the export timer then it'll retry the request again.
> 
> This makes sense to me because we don't know the size of each payload so we don't know if it'll take 4 seconds for 8 seconds for the request to complete. So if a request with a large payload takes longer to complete, we should let the request do that within the allowed export time instead of aborting the request early (ex. before the 10 seconds mark).
> 
> Do you agrees with this? @dyladan @blumamir @Flarna

This certainly makes sense for some cases where the payload is simply taking too long, but that would be quite a massive payload or quite a slow connection and I would expect the end-user to modify their timeout settings in that case.

There are also cases where this would prevent a potentially helpful retry. Intermittent network hiccups can cause a connection to stall when creating a new one would be faster (example you were routed to a bad network router or load balancer on the first try). In these cases you would want to retry. I think in the end it will most likely make sense to pick reasonable defaults that work for 99.9% of cases and allow the unusual user to configure them if need be.

svetlanabrennan commented:
Got it. Thanks for the update. I'll update this pr to mimic the java/.Net implementation of it.

svetlanabrennan commented:
I added tests to grpc & http (node) exporter. I wasn't able to add tests (yet) to the browser exporter to test `xmlhttprequest` timeout due to some error with "webpack".

Example of error:

```
  Uncaught ReferenceError: process is not defined
  at test/browser/index-webpack.js:39912:18
  
  ReferenceError: process is not defined
  at Object.<anonymous> (test/browser/index-webpack.js:39912:18)
      at Object.../../node_modules/stream-http/node_modules/readable-stream/lib/_stream_writable.js (test/browser/index-webpack.js:40543:30)
      ...
 ```
 
 I tried to run browser tests on a few different node versions (12, 14, 16) and I get the same error for all of them. 
 
 Anyone familiar with this error?

legendecas commented:
To my first view it seems like the problem is referencing "process" in the browser context: https://github.com/open-telemetry/opentelemetry-js/pull/2738/files#diff-2922953f04ebf0d689ee6cd427a51d486de3135179bf887240324dd8adc97366R66. You can get a platform-agnostic environment API from @opentelemetry/core https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-core/src/platform/node/globalThis.ts.

legendecas commented on packages/exporter-trace-otlp-http/src/util.ts:
```diff
@@ -39,3 +42,44 @@ export function appendResourcePathToUrlIfNotPresent(url: string, path: string):
+
+function getExporterTimeoutFromEnv(): number {
+  const definedTimeout =
+    Number(process.env.OTEL_EXPORTER_OTLP_TRACES_TIMEOUT ||
```
```suggestion
    Number(getEnv().OTEL_EXPORTER_OTLP_TRACES_TIMEOUT ||
      getEnv().OTEL_EXPORTER_OTLP_TIMEOUT);
```

svetlanabrennan commented:
This PR is now ready for review. Sorry about the delay.

legendecas commented:
The CI failures could be fixed by https://github.com/open-telemetry/opentelemetry-js/pull/2788.

svetlanabrennan commented:
It looks like there are some tests failing now because of the latest main merge into this branch. Hmmm.

svetlanabrennan commented:
Tests have been fixed. 

Most of the tests that were calling `.export` were failing because they weren't finishing the request. This became a problem with the addition of the `exporterTimer` in the `sendWithXhr` and `sendWithHttp`. This `exporterTimer` gets cleared when the request is either successful or not or errored which wasn't happening since the requests weren't finishing in many tests. 

This meant I had to add a mocked response to finish the requests in order to clear the `exporterTimer`. [Here's an example of this](https://github.com/open-telemetry/opentelemetry-js/blob/e8db3dca82cc6ccb51ecedfa9de57e961df0c2a1/packages/exporter-trace-otlp-http/test/node/CollectorTraceExporter.test.ts#L174-L178).

MSNev commented on packages/exporter-trace-otlp-http/src/OTLPExporterBase.ts:
```diff
@@ -36,6 +37,7 @@ export abstract class OTLPExporterBase<
   protected _concurrencyLimit: number;
   protected _sendingPromises: Promise<unknown>[] = [];
   protected _shutdownOnce: BindOnceFuture<void>;
+  public readonly _timeoutMillis: number;
```
Should public members be preceeded by an underscore "_" ?

Flarna replied:
public properties usually have no `_` prefix. I would also prefer to move it up a few lines to the other public properties.

Flarna commented on packages/exporter-trace-otlp-grpc/README.md:
````diff
@@ -107,6 +107,31 @@ provider.register();
++ Provide `timeoutMillis` to OTLPTraceExporter with `collectorOptions`:
+
+  ```js
+  const collectorOptions = {
````
This looks a bit confusing to create only a `const collectorOptions`  to configure an Exporter.

Should this be `exporterOptions`?  And maybe add a line to pass this options to a constructor like this:
```js
const exporterConfig = {
  timeoutMillis: 15000,
  // other options
};
const exporter = new OTLPTraceExporter(exporterConfig);
```

similar in http readme

MSNev commented on packages/exporter-trace-otlp-http/src/util.ts:
```diff
@@ -39,3 +42,44 @@ export function appendResourcePathToUrlIfNotPresent(url: string, path: string):
+    Number(getEnv().OTEL_EXPORTER_OTLP_TRACES_TIMEOUT ||
+    getEnv().OTEL_EXPORTER_OTLP_TIMEOUT);
+
+  if (definedTimeout) {
```
This will be false if the value is zero (0), so the next line <= 0 is really only < 0

Flarna commented on packages/exporter-trace-otlp-http/src/platform/node/util.ts:
```diff
@@ -65,21 +72,30 @@ export function sendWithHttp<ExportItem, ServiceRequest>(
     res.on('end', () => {
       if (res.statusCode && res.statusCode < 299) {
         diag.debug(`statusCode: ${res.statusCode}`, responseData);
+        clearTimeout(exporterTimer);
```
you could move this statement outside of the if to avoid duplication.

Flarna commented on packages/exporter-trace-otlp-http/src/util.ts:
```diff
@@ -39,3 +42,44 @@ export function appendResourcePathToUrlIfNotPresent(url: string, path: string):
+
+function getExporterTimeoutFromEnv(): number {
+  const definedTimeout =
+    Number(getEnv().OTEL_EXPORTER_OTLP_TRACES_TIMEOUT ||
```
```suggestion
    Number(getEnv().OTEL_EXPORTER_OTLP_TRACES_TIMEOUT ??
```

Flarna commented on packages/exporter-trace-otlp-http/test/browser/CollectorTraceExporter.test.ts:
```diff
@@ -243,28 +253,39 @@ describe('OTLPTraceExporter - web', () => {
     describe('when "sendBeacon" is available', () => {
+      let clock: sinon.SinonFakeTimers;
       beforeEach(() => {
+         // fakeTimers is used to replace the next setTimeout which is
```
```suggestion
        // fakeTimers is used to replace the next setTimeout which is
```

MSNev commented on packages/exporter-trace-otlp-http/src/platform/browser/util.ts:
```diff
@@ -51,9 +51,17 @@ export function sendWithXhr(
+  }, exporterTimeout);
+
   const xhr = new XMLHttpRequest();
   xhr.open('POST', url);
```
As long as you are not performing a synchronous request, XHR has a `timeout` field. Why not just ust that and then listen to the onTimeout?

Flarna commented on packages/exporter-trace-otlp-http/test/node/CollectorTraceExporter.test.ts:
```diff
@@ -354,4 +409,44 @@ describe('OTLPTraceExporter - node with json over http', () => {
+      collectorExporter.export(spans, responseSpy);
+
+      setTimeout(() => {
+        fakeRequest.emit('error', { code: 'ECONNRESET'});
```
Why emit an error after 100ms which is exactly the request timeout? isn't this some sort of race?

MSNev commented on packages/exporter-trace-otlp-http/src/platform/node/util.ts:
```diff
@@ -42,7 +42,14 @@ export function sendWithHttp<ExportItem, ServiceRequest>(
   const parsedUrl = new url.URL(collector.url);
+  let reqIsDestroyed: boolean;
+
+  const exporterTimer = setTimeout(() => {
```
Add the timeout property to the options
```typescript
const options = {
    //...
    timeout: 3000,
};
```

Flarna commented on packages/exporter-trace-otlp-http/test/node/CollectorTraceExporter.test.ts:
```diff
@@ -354,4 +409,44 @@ describe('OTLPTraceExporter - node with json over http', () => {
+  describe('export - with timeout', () => {
+    beforeEach(() => {
+      fakeRequest = new Stream.PassThrough();
+      stubRequest = sinon.stub(http, 'request').returns(fakeRequest as any);
```
Not sure if it is possible in CI but I think some test using a real HTTP request would be good. Currently it seems we just stub HTTP so we might miss some HTTP special stuff.
There might be different behavior at which stage of the HTTP request the timeout happens (socket creation, sending request, receiving response,...).
There might be also differences between node versions.

dyladan replied:
Has this been addressed?

legendecas replied:
I think it is ok here. Since the timer is created manually (https://github.com/open-telemetry/opentelemetry-js/pull/2738/files#diff-d5312ed9e76f7794f25948cf9ac39d6ee5103761f551fef74f8c154362e01d94R49) instead of the socket timer, it is just a timer that verifies that the request ends in a specific duration, otherwise we abort the request and log the event.

svetlanabrennan commented:
@Flarna Just checking in to see if you had any questions about my reply to your question above.

Flarna commented:
Which reply?

svetlanabrennan commented:
> 

[This one](https://github.com/open-telemetry/opentelemetry-js/pull/2738/files/4f9a5569ce85eb8502fd4d7611f68130c6faf4b3#r817933695) about using real http requests.

Flarna commented:
I can't see a reply from you there. Only my statement and a reply from @dyladan

svetlanabrennan commented:
That's weird. Well here's my reply to your question:

"Well does it matter which stage the http request timeouts in? The exporter timeout is used for the entire export process so it doesn't take into consideration what stage the request is in before it aborts the request, it just [aborts it.](https://nodejs.org/api/http.html#requestdestroyerror)

Can you clarify what HTTP special stuff would apply in this case (where the request is aborted in a certain number of milliseconds no matter what stage the request is in)?"

Flarna commented:
The [HTTP doc](https://nodejs.org/dist/latest-v16.x/docs/api/http.html#httprequesturl-options-callback) (scroll down a bit there) shows quite some different sequences of events signaled depending on when abort is called.
So there might be cases which are no covered by the stubbed http client.

But I'm not sure here, maybe everything is covered. It's maybe also quite hard to get all these cases in a stable way into a unitest.

legendecas commented on packages/exporter-trace-otlp-grpc/README.md:
```diff
@@ -115,9 +145,9 @@ Note, that this will only work if TLS is also configured on the server.
-- For more information on OpenTelemetry, visit: <https://opentelemetry.io/>
-- For more about OpenTelemetry JavaScript: <https://github.com/open-telemetry/opentelemetry-js>
-- For help or feedback on this project, join us in [GitHub Discussions][discussions-url]
++ For more information on OpenTelemetry, visit: <https://opentelemetry.io/>
```
Out of curiosity, what's the benefit of this change?

legendecas commented on packages/exporter-trace-otlp-grpc/src/util.ts:
```diff
@@ -84,10 +84,12 @@ export function send<ExportItem, ServiceRequest>(
 ): void {
   if (collector.serviceClient) {
     const serviceRequest = collector.convert(objects);
+    const deadline = new Date().setSeconds(new Date().getSeconds() + (collector.timeoutMillis / 1000));
```
Nit: can we calculate the deadline using unix-epoch instead of using the implicit-coercing setter of Date? gRPC client supports milliseconds as the deadline value: https://grpc.github.io/grpc/node/grpc.html#~Deadline

legendecas commented on packages/exporter-trace-otlp-proto/README.md:
```diff
@@ -38,6 +38,37 @@ provider.register();
+    }, //an optional object containing custom headers to be sent with each request will only work with http
+  };
+
+  const exporter = new OTLPTraceExporter(collectorOptions);
```
Do we have any support of these environs in package exporter-trace-otlp-proto? I didn't find any code changes.

svetlanabrennan commented:
> The [HTTP doc](https://nodejs.org/dist/latest-v16.x/docs/api/http.html#httprequesturl-options-callback) (scroll down a bit there) shows quite some different sequences of events signaled depending on when abort is called. So there might be cases which are no covered by the stubbed http client. Besides that node streams (used in http request/respons object) have seen quite some changes between node 8 and 17. So different node versions may show different behavior.

Thanks for the feedback on this!

I've been reviewing the sequence of events that are signaled depending on when using `req.destroy()` (node v17-14) or `req.abort()` (node v13-8) and found that the events are the same for these two cases:

- "request destroyed/aborted before a socket is assigned":
   - v17-13: `'error'` event emitted with an error with message `'Error: socket hang up'` and code `'ECONNRESET'`
   - v12-8:  documentation doesn't provide details on the events emitted for this use case for these versions but it appears that it shows `'error'` event emitted with an error with message `'Error: socket hang up'` and code `'ECONNRESET'` when I tested this use case locally
   
- "request destroyed/aborted before the connection succeeds":
   - v17-8: 'error' event  emitted with an error with message `'Error: socket hang up'` and code `'ECONNRESET'`

However, for this use case - "request destroyed/aborted after the response is received":
   - there is no `error` event emitted so I'm checking on how we can return a `Request Timeout` in this case that will work for all node v17-8.

dyladan commented:
> > The [HTTP doc](https://nodejs.org/dist/latest-v16.x/docs/api/http.html#httprequesturl-options-callback) (scroll down a bit there) shows quite some different sequences of events signaled depending on when abort is called. So there might be cases which are no covered by the stubbed http client. Besides that node streams (used in http request/respons object) have seen quite some changes between node 8 and 17. So different node versions may show different behavior.
> 
> Thanks for the feedback on this @Flarna
> 
> I've been reviewing the sequence of events that are signaled depending on when using `req.destroy()` (node v17-14) or `req.abort()` (node v13-8) and found that the events are the same for these two cases:
> 
> * "request destroyed/aborted before a socket is assigned":
>   
>   * v17-13: `'error'` event emitted with an error with message `'Error: socket hang up'` and code `'ECONNRESET'`
>   * v12-8:  documentation doesn't provide details on the events emitted for this use case for these versions but it appears that it shows `'error'` event emitted with an error with message `'Error: socket hang up'` and code `'ECONNRESET'` when I tested this use case locally
> * "request destroyed/aborted before the connection succeeds":
>   
>   * v17-8: 'error' event  emitted with an error with message `'Error: socket hang up'` and code `'ECONNRESET'`
> 
> However, for this use case - "request destroyed/aborted after the response is received":
> 
> * there is no `error` event emitted so I'm checking on how we can return a `Request Timeout` in this case that will work for all node v17-8.

I think his point was that things change even with minor versions and bugs could cause the documentation to be incorrect. The best way to ensure we are actually correct is to test with a real HTTP call somehow (possibly using nock to intercept?)

Flarna commented:
I don't think we should invest too much time into tests using real HTTP calls if it turns out to be complicated or flaky. I just wanted to point out that HTTP requests are quite a complicate state machinery and simply mocking at high level to fit one usecase is a bit risky.
We can always improve this later in case we are flooded with bug reports.

svetlanabrennan commented:
I updated the logic around destroying/aborting the request in `sendWithHttp` and a few simple real http requests:

If a request is destroyed/aborted after the response is received, [this code](https://github.com/open-telemetry/opentelemetry-js/blob/d955eac28a7c8b7dd9b546e3599da4a54f54dcf3/packages/exporter-trace-otlp-http/src/platform/node/util.ts#L78-L85) will handle returning a "request timeout" error.  This `'aborted'` event is emitted on `res` in node v8-17. 

And [this code](https://github.com/open-telemetry/opentelemetry-js/blob/d955eac28a7c8b7dd9b546e3599da4a54f54dcf3/packages/exporter-trace-otlp-http/src/platform/node/util.ts#L109-L113) continues to handle the case where a request is destroyed/aborted before a socket is assigned and before a connection succeeds. An `'error'` on `req` is emitted in this case for node v8-17.

I added a few real http request tests for the [http exporter](https://github.com/open-telemetry/opentelemetry-js/blob/d955eac28a7c8b7dd9b546e3599da4a54f54dcf3/packages/exporter-trace-otlp-http/test/node/CollectorTraceExporter.test.ts#L452-L528) and the [proto exporter](https://github.com/open-telemetry/opentelemetry-js/blob/d955eac28a7c8b7dd9b546e3599da4a54f54dcf3/packages/exporter-trace-otlp-proto/test/OTLPTraceExporter.test.ts#L286-L362). @Flarna Is this what you were looking for?

Flarna commented on packages/exporter-trace-otlp-http/src/platform/node/util.ts:
```diff
@@ -62,24 +74,47 @@ export function sendWithHttp<ExportItem, ServiceRequest>(
+    res.on('end', () => {
+      // node version <= 12 will emit 'end' event even if request was aborted
+      // no need to handle promise here if request was aborted since promise
+      // was already rejected in res.on('aborted')
```
The comment is not clear to me. It mentions a version diff for <= 12 but it seems there is no if checking the node version.

Flarna commented on packages/exporter-trace-otlp-proto/test/OTLPTraceExporter.test.ts:
```diff
@@ -261,7 +277,86 @@ describe('OTLPTraceExporter - node with proto over http', () => {
+      server.listen(8080);
+    });
+    afterEach(() => {
+      server.close();
```
[`server.close()` is async](https://nodejs.org/dist/latest-v16.x/docs/api/http.html#serverclosecallback) so maybe use
```js
afterEach((done) => {
      server.close(done);
}
```
```

Flarna commented on packages/exporter-trace-otlp-http/src/platform/node/util.ts:
```diff
@@ -42,7 +42,19 @@ export function sendWithHttp<ExportItem, ServiceRequest>(
+
+  const exporterTimer = setTimeout(() => {
+    reqIsDestroyed = true;
+    if (Number(nodeVersion) >= 14) {
```
```suggestion
    if (nodeVersion >= 14) {
```
No need for conversion here as it's already done during parsing

svetlanabrennan commented:
Update on status on this pr: waiting on [this pr](https://github.com/open-telemetry/opentelemetry-js/pull/2892) before I can make final changes proposed by @Flarna.

svetlanabrennan commented:
Ready for review.

Flarna commented:
@svetlanabrennan Lint is unhappy after merging master because intents are enforced now
```
 /home/runner/work/opentelemetry-js/opentelemetry-js/experimental/packages/otlp-exporter-base/test/browser/util.test.ts
Error:   58:1  error  Expected indentation of 10 spaces but found 8  indent
Error:   59:1  error  Expected indentation of 8 spaces but found 6   indent
```

svetlanabrennan commented:
> @svetlanabrennan Lint is unhappy after merging master because intents are enforced now
> 
> ```
>  /home/runner/work/opentelemetry-js/opentelemetry-js/experimental/packages/otlp-exporter-base/test/browser/util.test.ts
> Error:   58:1  error  Expected indentation of 10 spaces but found 8  indent
> Error:   59:1  error  Expected indentation of 8 spaces but found 6   indent
> ```

Ah. I'll get it fixed now.

svetlanabrennan commented:
Lint test is passing now. The codecov tests are failing but they were passing previously.  The codecov path test is saying that the [`sendWithHttp` function isn't being test](https://github.com/open-telemetry/opentelemetry-js/blob/e81f9dd458be66975d090dc42ea7599dd12eb3dd/experimental/packages/otlp-exporter-base/src/platform/node/util.ts#L45-L111) but it is being tested - the [tests are just located](https://github.com/open-telemetry/opentelemetry-js/blob/e81f9dd458be66975d090dc42ea7599dd12eb3dd/experimental/packages/exporter-trace-otlp-http/test/node/CollectorTraceExporter.test.ts#L466-L541) in another package. 

Should I move [these tests](https://github.com/open-telemetry/opentelemetry-js/blob/e81f9dd458be66975d090dc42ea7599dd12eb3dd/experimental/packages/exporter-trace-otlp-http/test/node/CollectorTraceExporter.test.ts#L466-L541) into the `otlp-exporter-base` package?

svetlanabrennan commented:
Tests failing after main merge. I'm seeing this test failing in the 'main' branch as well (when I run the tests locally).

Flarna commented:
@dyladan moved the latet label of @opentelemetry/api from verion 1.04 to 1.1.0. This could be the cause here in case several API versions are installed.

Flarna commented:
local tests on main are green for me. Seems like CI specific failures. Maybe some caching?

svetlanabrennan commented:
It must have been the cache. The local tests are passing now both in main and in this pr branch.

svetlanabrennan commented:
> @dyladan moved the latest label of @opentelemetry/api from verion 1.04 to 1.1.0. This could be the cause here in case several API versions are installed.

Was this a codecov config change? I assume so because these tests are all passing when I run it locally.

dyladan commented:
Codecov would not affect the tests anyway but this is probably because the 1.1 api is now latest. There is most likely a version mismatch somewhere. I'm on my phone and ooo right now but may be able to look later if I have time

svetlanabrennan commented:
Looks like 1 browser test is failing after merging main into this branch. Not sure why. All tests passing when I run it locally.

Flarna commented:
re-triggered the test and it's green now. Seems to be a test issue

svetlanabrennan commented:
I think there's also a test issue with the codecov for "codecov/patch" because it's indicating that some lines are not covered by tests but they are tested. 

For example, it says line 54 isn't tested in [this file](https://github.com/open-telemetry/opentelemetry-js/blob/fbaa4574d10c3f1bbc1fbeaedb63eb67633556b4/experimental/packages/otlp-exporter-base/src/util.ts#L54) but there is a test for it [here](https://github.com/open-telemetry/opentelemetry-js/blob/fbaa4574d10c3f1bbc1fbeaedb63eb67633556b4/experimental/packages/otlp-exporter-base/test/node/util.test.ts#L34).

legendecas commented on experimental/packages/exporter-trace-otlp-http/test/browser/CollectorTraceExporter.test.ts:
```diff
@@ -14,21 +14,22 @@
 import * as sinon from 'sinon';
 import { OTLPTraceExporter } from '../../src/platform/browser/index';
-
+import { nextTick } from 'process';
```
This is a browser test and we should prevent using Node.js API in browser codes.

legendecas commented on experimental/packages/exporter-trace-otlp-http/test/browser/CollectorTraceExporter.test.ts:
```diff
@@ -191,7 +197,7 @@ describe('OTLPTraceExporter - web', () => {
         });
 
-        setTimeout(() => {
+        nextTick(() => {
```
Ditto. Probably we can use `queueMicrotask` here.

dyladan replied:
setTimeout with a time of 0 would probably also work

legendecas reviewed (approved):
LGTM % nits

dyladan commented on experimental/CHANGELOG.md:
```diff
@@ -60,6 +60,7 @@ All notable changes to experimental packages in this project will be documented
 * feat(proto): add @opentelemetry/otlp-transformer package with hand-rolled transformation #2746 @dyladan
 * feat(sdk-metrics-base): shutdown and forceflush on MeterProvider #2890 @legendecas
 * feat(sdk-metrics-base): return the same meter for identical input to getMeter #2901 @legendecas
+* feat(otlp-exporter): add timeout env var #2738 @svetlanabrennan
```
can you add the specific env var to the changelog? bonus points if you can link to its documentation

dyladan reviewed (approved):
LGTM % legendecas' nit

svetlanabrennan commented:
One test failing in an unrelated package (didn't work on that package in this pr). 

Could this be a test issue again like [@Flarna mentioned earlier](https://github.com/open-telemetry/opentelemetry-js/pull/2738#issuecomment-1118985193) for another time an unrelated test was failing?

svetlanabrennan commented:
> Failure seems to be in the OTLPTraceExporter which is touched by this PR

The real http request tests were failing randomly. I think it was a timing issue. So I moved those tests to their own test suite and it's passing now. Also ran it locally a handful of times and it's all green as well.

dyladan commented on experimental/packages/otlp-exporter-base/src/platform/browser/OTLPExporterBrowserBase.ts:
```diff
@@ -78,7 +78,7 @@ export abstract class OTLPExporterBrowserBase<
-        sendWithXhr(body, this.url, this._headers, resolve, reject);
+        sendWithXhr(body, this.url, this._headers, this.timeoutMillis, resolve, reject);
       } else {
         sendWithBeacon(body, this.url, { type: 'application/json' }, resolve, reject);
```
beacon send does not need a timeout?

dyladan commented on experimental/packages/otlp-exporter-base/src/platform/node/util.ts:
```diff
@@ -42,7 +42,19 @@ export function sendWithHttp<ExportItem, ServiceRequest>(
+
+  const exporterTimer = setTimeout(() => {
+    reqIsDestroyed = true;
+    if (nodeVersion >= 14) {
+      req.destroy();
+    } else {
+      req.abort();
```
Can we add a comment why this needs to be different for 2 node versions?

dyladan reviewed (commented):
Overall LGTM was just a bit surprised to see the behavior split by major node version
